src/mopidy/internal/jsonrpc.py

```diff
@@ -1,59 +1,125 @@
 import inspect
-import json
 import traceback
 from collections.abc import Callable
-from typing import Any, Literal, TypeAlias, TypedDict, TypeVar
+from typing import Any, Literal, TypeAlias, TypeVar
 
+import pydantic_core
 import pykka
+from pydantic import (
+    BaseModel,
+    ConfigDict,
+    Field,
+    SerializerFunctionWrapHandler,
+    TypeAdapter,
+    field_serializer,
+    model_serializer,
+)
+from pydantic_core import PydanticUndefined, PydanticUndefinedType
+
+from mopidy import models
 
 T = TypeVar("T")
 
 
-class Notification(TypedDict):
-    jsonrpc: Literal["2.0"]
-    method: str
-    params: list[Any] | dict[str, Any]
+class UnsetType:
+    pass
+
+
+Unset = UnsetType()
 
 
 RequestId: TypeAlias = str | int | float
+Param: TypeAlias = (
+    # TODO: What types is needed by the core API?
+    models.Artist
+    | models.Album
+    | models.Track
+    | models.Playlist
+    | models.Ref
+    | models.Image
+    | Any
+)
+
+
+class Request(BaseModel):
```

Comment:
What's the difference between using `BaseModel` and `TypedDict` here?

Reply:
We must use pydantic's `BaseModel` to be able to do JSON deserialization using Pydantic, e.g. `Request.model_validate()`.

Reply:
But if `Request` is `TypedDict ` then can't you do:
```
request_adapter = TypeAdapter(Request)
my_validated_request = request_adapter.validate_json(json)
```
?

I find the pydantic docs unclear over which to use. And it says `TypedDict` is better [when nesting types](https://docs.pydantic.dev/latest/concepts/performance/#use-typeddict-over-nested-models) although my brain won't accept that without a bit of supporting explanation.

Reply:
Oh sorry, I'm being idiotic. We're trying to get our models out of this, a typed dict is still a dict. For some reason I thought it could be used like a `NamedTuple`. Sorry! Ignore!

Reply:
I think really I was barking down the pydantic.dataclasses tree. But they are not as good as Pydantic BaseModel, I get that.

Reply:
I think that using a `TypeAdapter` around a `TypedDict` where some of the fields are typed to be of one of our model types could work, so you're thinking is not that far out.